Hibernate-create-movie-session-hw

liudmylabiletska opened the pull request:
create addea CinemaHall.java, CinemaHallDao.java, CinemaHallDaoImpl.j…ava, CinemaHallService.java,CinemaHallServiceImpl.java, hibarnate.cfg.xml, Main.java, MovieSession.java, MovieSessionDao.java, MovieSessionDaoImpl.java, MovieSessionService.java, MovieSessionServiceImpl.java and pom.xml

liudmylabiletska committed:
create addea CinemaHall.java, CinemaHallDao.java, CinemaHallDaoImpl.java, CinemaHallService.java,CinemaHallServiceImpl.java, hibarnate.cfg.xml, Main.java, MovieSession.java, MovieSessionDao.java, MovieSessionDaoImpl.java, MovieSessionService.java, MovieSessionServiceImpl.java and pom.xml

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,14 +1,26 @@
 
 public class Main {
+    private static final Injector injector = Injector
+            .getInstance("mate.academy");
```
Good: injector instance is created using Injector.getInstance("mate.academy") which follows the task requirement to use the annotation-based injector. Keep this as-is so all services are retrieved from the injector rather than manually instantiated .

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,14 +1,26 @@
     public static void main(String[] args) {
-        MovieService movieService = null;
+        MovieService movieService = (MovieService) injector
+                .getInstance(MovieService.class);
```
You obtain MovieService from the injector here — that is correct. Note: casting is required because injector returns Object; this is fine. However, ensure MovieDaoImpl.getAll() exists and works (Main later calls movieService.getAll()). The task required implementing MovieDaoImpl.getAll() in DAO layer and MovieServiceImpl.getAll(); MovieServiceImpl.getAll is present but MovieDaoImpl.getAll must be implemented separately (not in this file) .

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,14 +1,26 @@
+        MovieService movieService = (MovieService) injector
+                .getInstance(MovieService.class);
+        CinemaHallService cinemaHallService = (CinemaHallService) injector
+                .getInstance(CinemaHallService.class);
```
You obtain CinemaHallService via the injector but you never call any of its methods. The description requires testing CinemaHallService methods (add, get, getAll) from main. Please add code that creates a CinemaHall, calls cinemaHallService.add(hall), cinemaHallService.get(hall.getId()), and cinemaHallService.getAll() and prints results so the service is actually exercised .

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,14 +1,26 @@
+        CinemaHallService cinemaHallService = (CinemaHallService) injector
+                .getInstance(CinemaHallService.class);
+        MovieSessionService movieSessionService = (MovieSessionService) injector
+                .getInstance(MovieSessionService.class);
```
You obtain MovieSessionService via the injector but you never call any of its methods. The task requires testing MovieSessionService methods (add, get, findAvailableSessions) from main. Add code to create a MovieSession (using a persisted Movie and CinemaHall and a LocalDateTime), call movieSessionService.add(...), get(...), and findAvailableSessions(movieId, date) and print outputs to verify behavior .

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,14 +1,26 @@
-        fastAndFurious.setDescription("An action film about street racing, heists, and spies.");
+        fastAndFurious.setDescription("An action film about "
+                + "street racing, heists, and spies.");
         movieService.add(fastAndFurious);
```
This code creates and adds a Movie and calls movieService.get and movieService.getAll, which is correct. Make sure the DAO implementation for Movie.getAll() exists and returns all movies; the service here delegates to movieDao.getAll() so DAO-side implementation is required for the call to succeed (implement MovieDaoImpl.getAll() if not done yet) .

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,14 +1,26 @@
 
         Movie fastAndFurious = new Movie("Fast and Furious");
-        fastAndFurious.setDescription("An action film about street racing, heists, and spies.");
+        fastAndFurious.setDescription("An action film about "
+                + "street racing, heists, and spies.");
```
Minor: you set the movie description across two concatenated string literals — that's fine. No action required here other than adding the missing service tests described above.

mateacademy-ai-mentor commented on src/main/java/mate/academy/model/CinemaHall.java:
```diff
@@ -0,0 +1,58 @@
+package mate.academy.model;
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Table;
```
Inconsistent JPA API: this file uses javax.persistence imports. Other entity classes in the project (for example MovieSession) use jakarta.persistence. Mixing javax and jakarta annotations can cause runtime mapping issues. Standardize on one API (prefer changing these imports to jakarta.persistence to match MovieSession) .

mateacademy-ai-mentor commented on src/main/java/mate/academy/model/CinemaHall.java:
```diff
@@ -0,0 +1,58 @@
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "cinema_holl")
```
Typo in table name: @Table(name = "cinema_holl") likely should be "cinema_hall". This mismatch will make Hibernate expect/create the wrong DB table. Fix the table name or ensure your DB uses this exact name intentionally.

mateacademy-ai-mentor commented on src/main/java/mate/academy/model/CinemaHall.java:
```diff
@@ -0,0 +1,58 @@
+    private int capacity;
+    private String description;
+
+    public CinemaHall() {
+    }
```
No-arg constructor visibility: JPA requires a no-arg constructor, but it can be protected instead of public. Consider changing `public CinemaHall()` to `protected CinemaHall()` to restrict visibility while still satisfying Hibernate. If you rely on frameworks that require public constructors, keep it public — this is advisory.

mateacademy-ai-mentor commented on src/main/java/mate/academy/model/MovieSession.java:
```diff
@@ -0,0 +1,79 @@
+package mate.academy.model;
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
```
This class uses jakarta.persistence imports. The project's CinemaHall entity currently uses javax.persistence imports; mixing jakarta.* and javax.* JPA annotations across entities can cause runtime mapping issues. Make JPA imports consistent across all entities (prefer jakarta.persistence if your project uses Jakarta API) and then re-run the application to verify mappings work. See CinemaHall for the inconsistent imports and table name there.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I’m requesting changes because there are critical blockers that prevent the task from meeting core requirements and may break runtime behavior.

Critical issues to fix:
- Main.java (src/main/java/mate/academy/Main.java): You obtain `CinemaHallService` and `MovieSessionService` but never call their methods. The task explicitly requires testing all their methods from main. Please add code that:
  - Creates a `CinemaHall`, calls `cinemaHallService.add(hall)`, then `cinemaHallService.get(hall.getId())` and `cinemaHallService.getAll()` and prints results.
  - Creates a `MovieSession` using a persisted `Movie` and the saved `CinemaHall` with a `LocalDateTime`, calls `movieSessionService.add(session)`, `movieSessionService.get(session.getId())`, and `movieSessionService.findAvailableSessions(movie.getId(), date)` and prints outputs.
- CinemaHall entity (src/main/java/mate/academy/model/CinemaHall.java): Inconsistent JPA imports vs MovieSession (javax vs jakarta). Standardize on one API (prefer `jakarta.persistence.*` to match MovieSession) to avoid mapping failures at runtime.
- CinemaHall entity (src/main/java/mate/academy/model/CinemaHall.java): Table name typo `@Table(name = "cinema_holl")`. Correct it to `cinema_hall` (or whatever your schema uses) to prevent Hibernate from targeting the wrong table.
- MovieDaoImpl.getAll(): `MovieServiceImpl.getAll()` delegates to `movieDao.getAll()`. Ensure `MovieDaoImpl.getAll()` is implemented; otherwise `Main` will fail when calling `movieService.getAll()`.

Minor (non-blocking) improvements:
- Consider making the no-arg constructor in `CinemaHall` protected instead of public.
- In `MovieSessionDaoImpl.findAvailableSessions`, you may use `date.atTime(23, 59, 59, 999_999_999)` for fully inclusive end-of-day matching (optional).

You’ve set up the injector usage and DAO/service layering well. Address the blockers above, then re-run to verify end-to-end behavior. Good progress—once these fixes are in, the solution should meet the requirements.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

liudmylabiletska committed:
create addea CinemaHall.java, CinemaHallDao.java, CinemaHallDaoImpl.java, CinemaHallService.java,CinemaHallServiceImpl.java, hibarnate.cfg.xml, Main.java, MovieSession.java, MovieSessionDao.java, MovieSessionDaoImpl.java, MovieSessionService.java, MovieSessionServiceImpl.java and pom.xml and modified CinemaHall.java, Main.java

mateacademy-ai-mentor commented on src/main/java/mate/academy/service/impl/MovieServiceImpl.java:
```diff
@@ -19,11 +19,12 @@ public Movie add(Movie movie) {
     @Override
     public Movie get(Long id) {
-        return movieDao.get(id).get();
+        return movieDao.get(id).orElseThrow(
+                () -> new RuntimeException("Movie with id " + id + " not found"));
```
The get(...) unwraps Optional and throws a RuntimeException when not found. This meets the service contract (return entity, not Optional). Consider using a domain-specific unchecked exception (e.g., NotFoundException or a project-level runtime) instead of raw RuntimeException to make errors explicit and testable.

Elena-Bruyako replied:
same here

mateacademy-ai-mentor commented on src/main/java/mate/academy/service/impl/MovieServiceImpl.java:
```diff
@@ -19,11 +19,12 @@ public Movie add(Movie movie) {
     @Override
     public List<Movie> getAll() {
-        return null;
+        return movieDao.getAll();
```
getAll() delegates to movieDao.getAll(). Ensure MovieDaoImpl.getAll() is implemented (session, query "FROM Movie", getResultList(), wrap exceptions in DataProcessingException). The project description requires adding getAll() to MovieDaoImpl; if that DAO method is missing the service will fail at runtime. See description requirement.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your service and DAO implementations are mostly solid and align with the checklist, but there are two critical blockers that will prevent the application from running end-to-end:

1) Missing `hibernate.cfg.xml` (critical, runtime blocker)
- Why: Without `src/main/resources/hibernate.cfg.xml`, Hibernate’s `SessionFactory` won’t initialize, so any DAO call that opens a session will fail at startup.
- How to fix: Add `src/main/resources/hibernate.cfg.xml` with DB connection (H2 in-memory is fine), dialect, `hbm2ddl.auto`, and entity packages/mappings. Verify `HibernateUtil.getSessionFactory()` initializes successfully.

2) Ensure `MovieDaoImpl.getAll()` is implemented (critical, runtime blocker)
- Why: `MovieServiceImpl.getAll()` delegates to `movieDao.getAll()`. If `MovieDaoImpl.getAll()` is missing, `Main` will fail at runtime when calling `movieService.getAll()`.
- How to fix: In `MovieDaoImpl`, implement `List<Movie> getAll()` using `session.createQuery("FROM Movie", Movie.class).getResultList()` inside try-with-resources, wrap exceptions in `DataProcessingException`, and annotate the class with `@Dao` so the injector picks it up.

Minor (non-blocking) suggestions:
- In `MovieServiceImpl#get(Long id)` (line with `orElseThrow`), prefer a domain-specific unchecked exception (e.g., `NotFoundException`) over `RuntimeException` to make not-found cases explicit and testable.
- Optional precision tweak in `MovieSessionDaoImpl.findAvailableSessions`: consider a half-open interval (`>= startOfDay AND < nextDayStart`) to include any sub-second times at day end.

You did well on: correct entity mappings, DAO patterns (transactions/try-with-resources), and service unwrapping of `Optional`. Once you add the config and the `MovieDaoImpl.getAll()` implementation, the app should run cleanly from `Main`. Keep up the good work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

liudmylabiletska committed:
create addea CinemaHall.java, CinemaHallDao.java, CinemaHallDaoImpl.java, CinemaHallService.java,CinemaHallServiceImpl.java, hibarnate.cfg.xml, Main.java, MovieSession.java,
MovieSessionDao.java, MovieSessionDaoImpl.java, MovieSessionService.java, MovieSessionServiceImpl.java and pom.xml and modified CinemaHall.java, Main.java modidifed hibernate.cfg.xml

liudmylabiletska committed:
create addea CinemaHall.java, CinemaHallDao.java, CinemaHallDaoImpl.java, CinemaHallService.java,CinemaHallServiceImpl.java, hibarnate.cfg.xml, Main.java, MovieSession.java,
MovieSessionDao.java, MovieSessionDaoImpl.java, MovieSessionService.java, MovieSessionServiceImpl.java and pom.xml and modified CinemaHall.java, Main.java modidifed hibernate.cfg.xml, modified MovieServiceImpl.java

liudmylabiletska committed:
create addea CinemaHall.java, CinemaHallDao.java, CinemaHallDaoImpl.java, CinemaHallService.java,CinemaHallServiceImpl.java, hibarnate.cfg.xml, Main.java, MovieSession.java,
MovieSessionDao.java, MovieSessionDaoImpl.java, MovieSessionService.java, MovieSessionServiceImpl.java and pom.xml and modified CinemaHall.java, Main.java modidifed hibernate.cfg.xml, modified MovieServiceImpl.java

liudmylabiletska committed:
modified MovieServiceImpl.java

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,16 +1,51 @@
 public class Main {
-    public static void main(String[] args) {
-        MovieService movieService = null;
+    private static final Injector injector = Injector.getInstance("mate.academy");
```
Runtime-blocker: The project is missing src/main/resources/hibernate.cfg.xml. Without it Hibernate's SessionFactory used by DAOs will not initialize and Main will fail at runtime. This violates the task requirement: "- create `src/main/resources/hibernate.cfg.xml`" . Add the hibernate.cfg.xml (H2 in-memory config is acceptable) and include mappings for Movie, CinemaHall and MovieSession.

mateacademy-ai-mentor commented on src/main/java/mate/academy/Main.java:
```diff
@@ -1,16 +1,51 @@
-        System.out.println(movieService.get(fastAndFurious.getId()));
+        System.out.println("Movie added: " + movieService.get(fastAndFurious
+                .getId()));
         movieService.getAll().forEach(System.out::println);
```
Runtime-blocker: This line calls movieService.getAll(), which delegates to the DAO. The task requires "- add implementations of `getAll()` method to `MovieDaoImpl` and `MovieServiceImpl'" — ensure MovieDaoImpl.getAll() is implemented (annotated @Dao) and returns a List<Movie> via a DB query (e.g. "FROM Movie"). If MovieDaoImpl.getAll() is missing the call here will fail at runtime .

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: moving you forward while highlighting a couple of runtime blockers to fix next.

What’s good
- Main.java correctly uses the Injector and exercises add/get/getAll/findAvailableSessions for Movie, CinemaHall, and MovieSession. This aligns well with the task’s testing flow.
- DAO/service patterns are solid overall: transactions, try/catch with DataProcessingException, and Optional unwrapping at the service layer look consistent and clean.

Critical issues to address
- Missing Hibernate config (src/main/resources/hibernate.cfg.xml): per requirement “create `src/main/resources/hibernate.cfg.xml`”. Without it, SessionFactory won’t initialize and Main will fail at runtime. Add H2 in-memory settings and ensure mappings for `Movie`, `CinemaHall`, `MovieSession` are included. Hint: verify `hibernate.dialect`, `jdbc.url`, `hbm2ddl.auto`, and annotated class mappings.
- MovieDaoImpl.getAll() implementation: Main calls `movieService.getAll()` which delegates to DAO. The requirement says to add `getAll()` to MovieDaoImpl; ensure `@Dao` is present and implement a query like `FROM Movie`, wrapping exceptions similar to other DAOs.

Minor improvement
- Consider a domain-specific unchecked exception for service “not found” cases instead of a generic RuntimeException.

You’ve structured the layers well—once the config file and MovieDaoImpl#getAll are in place, your Main should run smoothly. Keep going!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

Elena-Bruyako commented on src/main/java/mate/academy/model/CinemaHall.java:
```diff
@@ -0,0 +1,58 @@
+import jakarta.persistence.Table;
+
+@Entity
+@Table(name = "cinema_hall")
```
table name should be plural
```suggestion
@Table(name = "cinema_halls")
```

Elena-Bruyako commented on src/main/java/mate/academy/model/MovieSession.java:
```diff
@@ -0,0 +1,79 @@
+import java.time.LocalDateTime;
+
+@Entity
+@Table(name = "movie_session")
```
```suggestion
@Table(name = "movie_sessions")
```

Elena-Bruyako commented on src/main/java/mate/academy/service/impl/CinemaHallServiceImpl.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public CinemaHall get(Long id) {
+        return cinemaHallDao.get(id)
+                .orElseThrow(() -> new RuntimeException("Can't find cinema hall by id: " + id));
```
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can't find cinema hall by id: " + id));
```

Elena-Bruyako commented on src/main/java/mate/academy/service/impl/MovieSessionServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public MovieSession get(Long id) {
+        return movieSessionDao.get(id).orElseThrow(
+                () -> new RuntimeException("Movie session with id " + id + " not found"));
```
same here

Elena-Bruyako commented on pom.xml:
```diff
@@ -18,6 +18,11 @@
+            <groupId>javax.persistence</groupId>
+            <artifactId>javax.persistence-api</artifactId>
+            <version>2.2</version>
+        </dependency>
```
```suggestion
```

sarakhmen replied:
still not reverted. Please remove this dependency

Elena-Bruyako reviewed (changes requested):
see comments

liudmylabiletska committed:
modified MovieServiceImpl.java, CinemaHall.java, CinemaHallServiceImpl.java, MovieSession.java and MovieSessionServiceImpl.java

liudmylabiletska committed:
modified pom.xml
